### DISCUSSION TITLE
Can this tool obtain the namespace of each element in the C# file?

### DISCUSSION THREAD

**Discussion by yz-qiang** [dotnet/roslyn#65180] [Q&A]
Hi, Roslyn is a nice job, thanks for your contribution.
Here, I have a question about code analysis.
Given a `C#` code file, for example:

"using System;
namespace std
{
    class HelloWorld
    {
        static void Main(string[] args)
        {
            **Console.WriteLine("Hello World!");**
            **Console.ReadKey();**
        }}}
"

Whether Roslyn can  output a file that each code element be resolved into the fully-qualified name？Just Like

"
using System;
namespace std
{
    class HelloWorld
    {
        static void Main(string[] args)
        {
            **System.Console.WriteLine("Hello World!");**
            **System.Console.ReadKey();**
        }}}
“

If possible, can you show me how to implement this feature?
Thank you very much.

**Comment by jasonmalinowski** [comment]
@yz-qiang So yes, absolutely, that can be done. The one thing of course to keep in mind is you do need some project surrounding your source file for us to even know what "Console" is, since that comes from the framework.

Was there a specific thing you were having issues with?

**Reply by yz-qiang** [reply]
Thanks for your reply. 
First of all, the fully-qualified name resolution is a crucial step in our current research work.

I made some attempts according to the instructions at https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/get-started/syntax-transformation, but I failed.
If so, could you share the specific implementation code with me?
This step is vital in our current research work. Thank you very much.
Best yz-qiang.

**Reply by CyrusNajmabadi** [reply]
You'd have to provide more information on what failed. I also recommend using our discord for this (discord.gg/csharp) so you can more quickly converse with team members and the community about this.

**Reply by yz-qiang** [reply]
Hi,
Thanks for your reply.
At first, I want to get the fully-qualified name of each code element from the parsed AST.
However, when I used the following code, I failed to get the fully-qualified name from the pared AST.
`const string sampleCode = @"using System;
namespace std
{
class HelloWorld
{
static void Main(string[] args)
{
Console.WriteLine(""Hello, World!"");
Console.ReadKey();
}}}"
SyntaxTree tree = CSharpSyntaxTree.ParseText(sampleCode);
var root = (CompilationUnitSyntax)tree.GetRoot();
`
For example, I can't get the fully-qualified name of `Console.WriteLine` from the `root`, i.e., `System.Console.WriteLine()`.
I'm very eager to know how to achieve this target. I really need your help. Thank you very much.

**Reply by CyrusNajmabadi** [reply]
As above. I recommend discord. Right now you're just examining syntax. If your want semantics you have to use the Roslyn semantic apis. I recommend reading: https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/#next-steps

Especially the sections on syntax and semantics.

**Reply by yz-qiang** [reply]
Thanks. 
I read this documentation and ran examples successfully.
But I can't still achieve my goal.
Can you show me some examples, I desperately need them.
Thank you very much.
Best yz-qiang.

**Reply by CyrusNajmabadi** [reply]
What isn't working for you?

**Reply by yz-qiang** [reply]
Thanks.
Acutally, I don't know which APIs and how to implement this feature.
I would appreciate it if you could give me some ideas or concrete examples for implementing this feature.
Thank you very much.
Bert yz-qiang.

**Reply by CyrusNajmabadi** [reply]
Best docs would be: https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/get-started/semantic-analysis?source=recommendations

And I'd use discord as well to converse and ask more specific questions if your run into issues.

**Reply by yz-qiang** [reply]
Ok, Thanks.
Here is my written code:  （My question is in the notes. Thank you very much） 

```
using static System.Console;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
using System.Collections.Generic;
using System.Linq;
using System.CodeDom.Compiler;
using System.Dynamic;
using System;

namespace SyntaxWalker
{
    class Program
    {
        const string programText =
                @"using System;
                    using System.Collections.Generic;
                    using System.Text;
                    namespace HelloWorld
                    {
                        class Program
                        {
                            static void Main(string[] args)
                            {
                                Console.WriteLine(""Hello, World!"");
                            }
                        }
                    }";
        static void Main(string[] args)
        {
            SyntaxTree tree = CSharpSyntaxTree.ParseText(programText);
            CompilationUnitSyntax root = tree.GetCompilationUnitRoot();

            var collector = new Collector();
            collector.Visit(root);
            foreach (var eps in collector.EPS)
            {     
               //  **###### Here is my question #######**
               //  Here, I want to get the fully-qualified name of **Console.WriteLine**, that is **System.Console.WriteLine**. But I don't know which APIs can obtain this fully-qualified name.
                WriteLine(eps.Expression);  //  the result is  Console.WriteLine(""Hello, World!"");
            }
        }
    }
     class Collector : CSharpSyntaxWalker
    {
        public ICollection<ExpressionStatementSyntax>  EPS{ get; } = new List<ExpressionStatementSyntax>();
        public override void VisitExpressionStatement (ExpressionStatementSyntax node)
        {
            EPS.Add(node);
        }
    }
}
```

**Reply by CyrusNajmabadi** [reply]
See the docs that were linked. It's in this section: 
https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/get-started/semantic-analysis?source=recommendations#binding-a-name

I recommend using discord for this conversation. It will likely go much faster.

**Reply by yz-qiang** [reply]
Thanks for your reply.
According to your suggestion, I am trying to use the semantic model to achieve my goal.
However, I still have some questions. My code is the following (Note: **My question noted in the code.** ):
```
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Linq;
using System;

namespace SyntaxWalker
{
    class semanticModel
    {
        const string programText = @"using Microsoft.CodeAnalysis;
                                    using Microsoft.CodeAnalysis.CSharp;
                                    using Microsoft.CodeAnalysis.CSharp.Syntax;
                                    using System;
                                    using System.Linq;
                                    namespace ClassLibrary1 {
                                        class UsingCollector {
                                            const string programText = ...;
                                                static void Main(string[] args)
                                            {
                                                var tree = CSharpSyntaxTree.ParseText(programText);
                                                var Mscorlib = MetadataReference.CreateFromFile(typeof(object).Assembly.Location);
                                                var compilation = ....;
                                                var methodSyntax = tree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().First();
                                                Console.WriteLine(""Hello world"");
                                            }
                                        }
                                    }";
        static void Main(string[] args) {
            var tree = CSharpSyntaxTree.ParseText(programText);
            var Mscorlib = MetadataReference.CreateFromFile(typeof(object).Assembly.Location);
            var compilation = CSharpCompilation.Create("MyCompilation",
                    syntaxTrees: new[] { tree }, references: new[] { Mscorlib });
            var model = compilation.GetSemanticModel(tree);

            //Looking at the first method symbol
            var expression = tree.GetRoot().DescendantNodes().OfType<ExpressionStatementSyntax>().First();
            var expressionState = expression.Expression;
            var expressionSymbol = model.GetSymbolInfo(expressionState);
            var expressionSymbol_string = expressionSymbol.Symbol.ToString();

  // ① #### Here, the fully qualified name of `Console.WriteLine()' could be printed correctly, that is, System.Console.WriteLine()
            Console.WriteLine(expressionSymbol_string);  

            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().First();
            var invocationState = invocation.Expression;  // Here, the output is `CSharpSyntaxTree.ParseText(programText)`
            var invocationSymbol = model.GetSymbolInfo(invocationState);
            var invocationSymbol_string = invocationSymbol.Symbol.ToString();

             // ② #### However, the output is Null.  I don't know why.
             Console.WriteLine(invocationSymbol_string);

        }
    }
}
```
Otherwise, For ①, If I replace the `Console.WriteLine(""Hello world""); ` with `Console.WriteLine(methodSyntax.ToString()); `  in `programText`. The output of `expressionSymbol_string` will also be Null. It feels a little weird.

Intuitively, I find this approach a little complicated. Because first, I need to get the fully qualified name of a code element and then convert the simple name in the code to a fully qualified name. 
Alternatively: is there another (more suited?) way to do that?

**Reply by CyrusNajmabadi** [reply]
I recommend discord.

**Reply by yz-qiang** [reply]
Thanks for your reply.
What is discord? Could you tell me more about it？
And can you tell me what the problem is with my code?

**Reply by rafael-f** [reply]
This is Discord -> http://aka.ms/discord-csharp-roslyn

(Chat app)
